### ISSUE TITLE
[Android] NPE in BackgroundGeolocationModule.stopBackgroundService()

### ISSUE THREAD

**Issue by GuillaumeLachaud** [mauron85/react-native-background-geolocation#121]
## Your Environment
* Plugin version: 0.3.1
* Platform: Android
* OS version: Android 7.0 (API 24)
* Device manufacturer and model: Samsung SM-G955U
* React Native version: 0.49.0
* Plugin configuration options: 
 BackgroundGeolocation.configure({
        desiredAccuracy: 10,
        stationaryRadius: 50,
        distanceFilter: 50,
        locationTimeout: 30,
        notificationTitle: 'TITLE',
        notificationText: 'TEXT',
        debug: false,
        startOnBoot: false,
        stopOnTerminate: false,
        locationProvider: BackgroundGeolocation.provider.ANDROID_ACTIVITY_PROVIDER,
        interval: LOCATION_UPDATE_INTERVAL,
        fastestInterval: 10 * 1000,
        activitiesInterval: 10000,
        stopOnStillActivity: false,
        url: 'URL',
        httpHeaders: {userId}
    });
* Link to your project: N/A

## Context
App crashes when resumed after being killed from recent tasks.

## Expected Behavior
App doesn't crash

## Actual Behavior
App crashes

## Possible Fix
Use `ReactApplicationContext` rather than `currentActivity` for both `startBackgroundService()` and `stopBackgroundService()`

## Steps to Reproduce

1. Start app
2. Start location collection
3. Put app in background
4. Kill app from recent apps list
5. Reopen app

## Context

It looks like the weak reference to the current activity in `ReactContext` is garbage collected but the as the service keeps running in the background, any attempt to use the activity as a context will likely cause a NPE. Is there any reason we can't use the Application Context rather than an Activity? A Service doesn't have to be tied to an Activity.

## Debug logs
`android.content.ComponentName.<init> (ComponentName.java:128)
android.content.Intent.<init> (Intent.java:5390)
com.marianhello.react.BackgroundGeolocationModule.stopBackgroundService (BackgroundGeolocationModule.java:464)
com.marianhello.react.BackgroundGeolocationModule.stop (BackgroundGeolocationModule.java:289)
java.lang.reflect.Method.invoke (Method.java)
com.facebook.react.bridge.JavaMethodWrapper.invoke (JavaMethodWrapper.java:363)
com.facebook.react.bridge.JavaModuleWrapper.invoke (JavaModuleWrapper.java:162)
com.facebook.react.bridge.queue.NativeRunnable.run (NativeRunnable.java)
android.os.Handler.handleCallback (Handler.java:751)
android.os.Handler.dispatchMessage (Handler.java:95)
com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage (MessageQueueThreadHandler.java:31)
android.os.Looper.loop (Looper.java:154)
com.facebook.react.bridge.queue.MessageQueueThreadImpl$3.run (MessageQueueThreadImpl.java:194)
java.lang.Thread.run (Thread.java:762)
`

**Comment by mauron85** [comment]
Will take a look on this.

**Comment by mauron85** [comment]
The strange thing is that it crashed after app reopen.
On reopen `stopBackgroundService` should not be called at all. Is there any possibility you can stop () somewhere in your js init code?

**Comment by GuillaumeLachaud** [comment]
You're right, js code does call stop(). Here is a little bit more context:

This happens when the app is re-opened through a push notification. When this kind of push notification is received, my app needs to stop location collection. So it looks like this can happen when the push action, to stop location collection, is taken before the Activity is set in the `ReactContext` through `onHostResumed()`

**Comment by mauron85** [comment]
We can try your suggested solution with ReactContext.

Also I think you can eliminate your problem with upcoming version 0.4.

Simple add check if service is running with `checkStatus` method. If's not then do not call stop.

```
    BackgroundGeolocation.checkStatus(status => {
      console.log('[INFO] BackgroundGeolocation service is running', status.isRunning);
      console.log('[INFO] BackgroundGeolocation service has permissions', status.hasPermissions);
      console.log('[INFO] BackgroundGeolocation auth status: ' + status.authorization);

      // you don't need to check status before start (this is just the example)
      if (status.isRunning) {
        BackgroundGeolocation.stop();
      }
    });
```

**Comment by GuillaumeLachaud** [comment]
I can try but it looks like there is already a similar defensive check in `stopBackgroundService()`

```
protected void stopBackgroundService() {
        if (!mIsServiceRunning) { return; }

        log.info("Stopping bg service");
        final Activity currentActivity = this.getCurrentActivity();
        currentActivity.stopService(new Intent(currentActivity, LocationService.class));
        mIsServiceRunning = false;
    }
```

What will change in 0.4? Do you have any rough estimate of when it might get released?

Thanks!

**Comment by mauron85** [comment]
Things are little bit more complicated when you use `stopOnTerminate: false`.
When you terminate your app. Main activity obviously will be destroyed (but service if was running will be still running even without bind to Main Activity).

And now you reopen your app. BackgroundGeolocationModule lives in "context" of MainActivity so `mIsServiceRunning: false` eventhough sevice is actually running. There is no way in plugin version to check reliable is service is running when option `stopOnTerminate: false` is used.

Hopefully this explained what is happening your case. You simple call stop on on running service, which is unfortunately undefined behaviour (which I'm aware of).

I would say you can safely use v0.4 from `next` branch.

It's already in npm: `npm install react-native-mauron85-background-geolocation@next`
Docs have been adjusted also: https://github.com/mauron85/react-native-background-geolocation/tree/next

I will write down Breaking changes so I can copy them into docs later:

* start method doesn't accept callback (use .on("start") event
* stop methods doesn't accept callback (use .on("stop") event)
* for background syncing (of failed post) you now need to specify `syncUrl` option. In version 0.3.x if `syncUrl` was not set `url` was used.
* Android distance filter now accept arbitrary values (before only 10, 100, 1000) and map them to Android location criteria
* all plugin constants are in directly BackgroundGeolocation namespace. (check [index.js](https://github.com/mauron85/react-native-background-geolocation/blob/next/index.js))
* plugin can be started without executing `configure` (stored settings or defaults will be used)
* location property `locationId` renamed to just `id`
* iOS `pauseLocationUpdates` now default to false (becuase iOS docs now states that you need to restart manually if you set it to true)

I think only first two changes are major.

**Comment by mauron85** [comment]
I cannot reproduce your issue with steps you provided.

Tried on:
*Genymotion Samsung Galaxy 4.4 (emulator)
*Android x86 7.1 rc2 (emulator)
*CM 13.0 (Android 6.0) (device)

Testing on following App:
```
export default class App extends Component<{}> {
  componentWillMount() {
    BackgroundGeolocation.configure({
      desiredAccuracy: 10,
      stationaryRadius: 50,
      distanceFilter: 50,
      locationTimeout: 30,
      notificationTitle: 'TITLE',
      notificationText: 'TEXT',
      debug: false,
      startOnBoot: false,
      stopOnTerminate: false,
      locationProvider: BackgroundGeolocation.provider.ANDROID_DISTANCE_FILTER_PROVIDER,
      interval: 10 * 1000,
      fastestInterval: 10 * 1000,
      activitiesInterval: 10000,
      stopOnStillActivity: false,
      url: 'http://192.168.64.1:3000/locations',
      httpHeaders: { userId: '1' }
    });

    BackgroundGeolocation.on('location', (location) => {
      //handle your locations here
    });

    BackgroundGeolocation.on('stationary', (stationaryLocation) => {
      //handle stationary locations here
    });

    BackgroundGeolocation.on('error', (error) => {
      console.log('[ERROR] BackgroundGeolocation error:', error);
    });

    BackgroundGeolocation.stop();

    BackgroundGeolocation.start(() => {
      console.log('[DEBUG] BackgroundGeolocation started successfully');    
    });
  }
...
}
```

Can you post exact sequence of api calls that lead to crash?

**Comment by mauron85** [comment]
This should be already fixed in 0.5x branch.
